OSSM-8385: Installing multi-primary multi-network mesh (#85091)
openshift/openshift-docs: modules/ossm-verifying-multi-cluster-topology.adoc

// This procedure is used in the following assembly:
// * install/ossm-multi-cluster-topologies.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-verifying-multi-cluster-topology_{context}"]
= Verifying multi-cluster topology

> Deploy same applications and verify traffic on a multi-cluster topology on two {ocp-product-title} clusters. 

Reviewer: > Deploy **same** applications...

same or sample?

Author: Thanks!

In this procedure, `CLUSTER1` is the East cluster and `CLUSTER2` is the West cluster. 
====

.Prerequisites

* You have installed the {SMProduct} Operator on all of the clusters that comprise the mesh.

* You have completed "Creating certificates for a multi-cluster mesh". 

* You have completed "Applying certificates to a multi-cluster topology".

* You have created an {istio} Container Network Interface (CNI) resource.

* You have `istioctl` installed on the laptop you will use to run these instructions.

* You have installed a multi-cluster topology.

.Procedure

. Deploy sample applications on the East cluster:

.. Create a sample application namespace on the East cluster by running the following command:
+
[source,terminal]
----
$ oc --context "${CTX_CLUSTER1}" get project sample || oc --context="${CTX_CLUSTER1}" new-project sample
----

.. Label the application namespace to support sidecar injection by running the following command:
+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER1}" label namespace sample istio-injection=enabled
----

.. Deploy the `helloworld` application:

... Create the `helloworld` service by running the following command:
+
[source,terminal]
----

> oc --context="${CTX_CLUSTER1}" apply \

Reviewer: ```suggestion
$ oc --context="${CTX_CLUSTER1}" apply \
```

Author: done

  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/helloworld/helloworld.yaml \
  -l service=helloworld -n sample
----

... Create the `helloworld-v1` deployment by running the following command:
+
[source,terminal]
----

> oc --context="${CTX_CLUSTER1}" apply \

Reviewer: ```suggestion
$ oc --context="${CTX_CLUSTER1}" apply \
```

  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/helloworld/helloworld.yaml \
  -l version=v1 -n sample
----

.. Deploy the `sleep` application by running the following command:
+
[source,terminal]
----

> oc --context="${CTX_CLUSTER1}" apply \

Reviewer: ```suggestion
$ oc --context="${CTX_CLUSTER1}" apply \
```

  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/sleep/sleep.yaml -n sample
----

> . Wait for the sample applications on the East cluster to return the "Ready" status condition by running the following command:

Reviewer: This should be bullet 1.e, not 2.

Author: made this a sub-bullet

+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER1}" wait --for condition=available -n sample deployment/helloworld-v1

> $ oc --context="${CTX_CLUSTER1}" wait --for condition=available -n sample deployment/sleep

Reviewer: Two commands in the same block.

Author: broke apart

----

. Deploy the sample applications on the West cluster:

.. Create a sample application namespace on the West cluster by running the following command:
+
[source,terminal]
----
$ oc --context "${CTX_CLUSTER2}" get project sample || oc --context="${CTX_CLUSTER2}" new-project sample
----

.. Label the application namespace to support sidecar injection by running the following command:
+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER2}" label namespace sample istio-injection=enabled
----

.. Deploy the `helloworld` application:

... Create the `helloworld` service by running the following command:
+
[source,terminal]
----

> oc --context="${CTX_CLUSTER2}" apply \

Reviewer: ```suggestion
$ oc --context="${CTX_CLUSTER2}" apply \
```

Author: done

  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/helloworld/helloworld.yaml \
  -l service=helloworld -n sample
----

... Create the `helloworld-v2` deployment by running the following command:
+
[source,terminal]
----

> oc --context="${CTX_CLUSTER2}" apply \

Reviewer: ```suggestion
$ oc --context="${CTX_CLUSTER2}" apply \
```

  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/helloworld/helloworld.yaml \
  -l version=v2 -n sample
----

.. Deploy the `sleep` application by running the following command:
+
[source,terminal]
----

> oc --context="${CTX_CLUSTER2}" apply \

Author: done

  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/sleep/sleep.yaml -n sample
----

> . Wait for the sample applications on the West cluster to return the "Ready" status condition by running the following command:

Author: made this a sub-bullet

+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER2}" wait --for condition=available -n sample deployment/helloworld-v2
$ oc --context="${CTX_CLUSTER2}" wait --for condition=available -n sample deployment/sleep
----

> .Verifying traffic flows between clusters

Author: @luksa I broke this out as a separate procedure. Let's give this a brief discussion when we meet Tuesday.

> . For the East cluster, send 10 requests to the `helloworld` service by running the following command:

Author: done

+
[source,terminal]
----

> for i in {0..9}; do

Reviewer: ```suggestion
$ for i in {0..9}; do
```

Author: done

  oc --context="${CTX_CLUSTER1}" exec -n sample deploy/sleep -c sleep -- curl -sS helloworld.sample:5000/hello;
done
----
+
Verify that you see responses from both clusters. This means you will see version 1 and version 2 of the service in the responses. 

. For the West cluster, send 10 requests to the `helloworld` service:
+
[source,terminal]
----

> for i in {0..9}; do

Reviewer: ```suggestion
$ for i in {0..9}; do
```

Author: done

+
[source,terminal]
----
$ oc --context "${CTX_CLUSTER1}" get project sample || oc --context="${CTX_CLUSTER1}" new-project sample
----

.. Label the application namespace to support sidecar injection by running the following command:
+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER1}" label namespace sample istio-injection=enabled
----

.. Deploy the `helloworld` application:

... Create the `helloworld` service by running the following command:
+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER1}" apply \
  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/helloworld/helloworld.yaml \
  -l service=helloworld -n sample
----

... Create the `helloworld-v1` deployment by running the following command:
+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER1}" apply \
  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/helloworld/helloworld.yaml \
  -l version=v1 -n sample
----

.. Deploy the `sleep` application by running the following command:
+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER1}" apply \
  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/sleep/sleep.yaml -n sample
----

> .. Wait for the `helloworld` application on the East cluster to return the "Ready" status condition by running the following command:

Reviewer 2: ```suggestion
.. Wait for the `helloworld` application on the East cluster to return the `Ready` status condition by running the following command:
```

Author: done

  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/helloworld/helloworld.yaml \
  -l version=v2 -n sample
----

.. Deploy the `sleep` application by running the following command:
+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER2}" apply \
  -f https://raw.githubusercontent.com/istio/istio/${ISTIO_VERSION}/samples/sleep/sleep.yaml -n sample
----

.. Wait for the `helloworld` application on the West cluster to return the "Ready" status condition by running the following command:
+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER2}" wait --for condition=available -n sample deployment/helloworld-v2
----

.. Wait for the `sleep` application on the West cluster to return the "Ready" status condition by running the following command:
+
[source,terminal]
----
$ oc --context="${CTX_CLUSTER2}" wait --for condition=available -n sample deployment/sleep
----

.Verifying traffic flows between clusters

. For the East cluster, send 10 requests to the `helloworld` service by running the following command:
+
[source,terminal]
----
$ for i in {0..9}; do
  oc --context="${CTX_CLUSTER1}" exec -n sample deploy/sleep -c sleep -- curl -sS helloworld.sample:5000/hello;

> done

Author: @luksa  I think the slashes crashed the build, but I can try again later today.

----
+

> Verify that you see responses from both clusters. This means you will see version 1 and version 2 of the service in the responses. 

Reviewer 2: ```suggestion
Verify that you see responses from both clusters. This means version 1 and version 2 of the service can be seen in the responses. 
```

Author: done

. For the West cluster, send 10 requests to the `helloworld` service:
+
[source,terminal]
----
$ for i in {0..9}; do
  oc --context="${CTX_CLUSTER2}" exec -n sample deploy/sleep -c sleep -- curl -sS helloworld.sample:5000/hello;

> done

Author: done

----
+

> Verify that you see responses from both clusters. This means you will see version 1 and version 2 of the service in the responses. 

Reviewer 2: ```suggestion
Verify that you see responses from both clusters. This means version 1 and version 2 of the service can be seen in the responses. 
```

Reviewer 2: IBMSG suggests avoiding future tense.
